src/message_pool.zig

```diff
@@ -16,15 +16,48 @@ comptime {
 /// message to be shifted to make space for 0 padding to vsr.sector_ceil.
 const message_size_max_padded = config.message_size_max + config.sector_size;
 
-/// A pool of reference-counted Messages, memory for which is allocated only once
-/// during initialization and reused thereafter. The config.message_bus_messages_max
-/// and config.message_bus_headers_max values determine the size of this pool.
+/// The number of full-sized messages allocated at initialization by the replica message pool.
+/// There must be enough messages to ensure that the replica can always progress, to avoid deadlock.
+pub const messages_max_replica = messages_max: {
+    const client_table_messages_max = config.clients_max;
+    const journal_messages_max = config.io_depth_read + config.io_depth_write;
+    const loopback_queue_messages_max = 1;
+    const pipelining_messages_max = config.pipelining_max;
+    // Only 1 quorum stores messages (QuorumMessages): do_view_change_from_all_replicas.
+    const quorum_messages_max = config.replicas_max;
+
+    // +1 to account for the Connection's `recv_message`.
+    const connection_messages_max = config.connection_send_queue_max_replica + 1;
+    const message_bus_messages_max = config.connections_max * connection_messages_max;
+
+    break :messages_max pipelining_messages_max + quorum_messages_max +
+        loopback_queue_messages_max + client_table_messages_max + journal_messages_max +
+        message_bus_messages_max;
+};
+
+/// The number of full-sized messages allocated at initialization by the client message pool.
+pub const messages_max_client = messages_max: {
+    // +1 to account for the Connection's `recv_message`.
+    const connection_messages_max = config.connection_send_queue_max_client + 1;
+    const message_bus_messages_max = config.replicas_max * connection_messages_max;
+    // +1 to account for creating a ping when the send/request queues are already full.
+    break :messages_max 1 + message_bus_messages_max + config.client_request_queue_max;
```

Comment:
I did consider this case. But (and I admit this is an awfully obscure edge case) is there anything preventing both positions of the connection send queue from being (separate) pings?

Reply:
Perhaps we shouldn't try to get a ping message if the send queue is full? This might also be something to consider... not attempting to get a message if we see we can't ref it somewhere, because otherwise we might overflow our static allocation again.

Reply:
Although I think that's already covered by our `always add 1 to the total` principle to cover temporary overflow in the control plane.

Reply:
Yeah, I think "always add 1" means that it's safe (i.e. won't crash).

> Perhaps we shouldn't try to get a ping message if the send queue is full?

If I remember right, the client always pings all replicas at once (using a single many-reference ping message) so I think its probably simpler as-is, even though it can (rarely) create a ping that is immediately discarded by the message bus if every connection's send queue is full. It isn't much extra work, and a single message overhead in the static allocation is pretty minor (and is simpler than checking every send queue prior to the ping).

Despite the shared ping message, the client still needs to account for `replica_count` ping messages since they can send at different speeds (again, very much an obscure edge case).